proxy/pkg/zdmproxy/frameprocessor.go

```diff
@@ -0,0 +1,129 @@
+	return &streamIdProcessor{
+		mapper:   mapper,
+		connType: connType,
+		metrics:  metrics,
+	}
+}
+
+func (sip *streamIdProcessor) AssignUniqueId(rawFrame *frame.RawFrame) (*frame.RawFrame, error) {
+	if rawFrame == nil {
+		return rawFrame, nil
+	}
+	var newId, err = sip.getNewStreamId(rawFrame.Header.StreamId)
+	if err != nil {
+		return rawFrame, err
+	}
+	return setRawFrameStreamId(rawFrame, newId), nil
+}
+
+func (sip *streamIdProcessor) AssignUniqueIdFrame(frame *frame.Frame) (*frame.Frame, error) {
+	if frame == nil {
+		return frame, nil
+	}
+	var newId, err = sip.getNewStreamId(frame.Header.StreamId)
+	if err != nil {
+		return frame, err
+	}
+	return setFrameStreamId(frame, newId), nil
+}
+
+func (sip *streamIdProcessor) ReleaseId(rawFrame *frame.RawFrame) (*frame.RawFrame, error) {
+	if rawFrame == nil {
+		return rawFrame, nil
+	}
+	var originalId, err = sip.releaseStreamId(rawFrame.Header.StreamId)
+	if err != nil {
+		return rawFrame, err
+	}
+	return setRawFrameStreamId(rawFrame, originalId), err
+}
+
+func (sip *streamIdProcessor) ReleaseIdFrame(frame *frame.Frame) (*frame.Frame, error) {
+	if frame == nil {
+		return frame, nil
+	}
+	var originalId, err = sip.releaseStreamId(frame.Header.StreamId)
+	if err != nil {
+		return frame, err
+	}
+	return setFrameStreamId(frame, originalId), err
+}
+
+func setRawFrameStreamId(f *frame.RawFrame, id int16) *frame.RawFrame {
+	// If the new id is the same as the original id (most likely an internal request), then just return the original
+	// frame
+	if f.Header.StreamId == id {
+		return f
+	}
+	newHeader := f.Header.Clone()
+	newHeader.StreamId = id
+	return &frame.RawFrame{
+		Header: newHeader,
+		Body:   f.Body,
+	}
+}
+
+func setFrameStreamId(f *frame.Frame, id int16) *frame.Frame {
+	// If the new id is the same as the original id (most likely an internal request), then just return the original
+	// frame
+	if f.Header.StreamId == id {
+		return f
+	}
+	newHeader := f.Header.Clone()
+	newHeader.StreamId = id
+	return &frame.Frame{
+		Header: newHeader,
+		Body:   f.Body,
+	}
+}
+
+// getNewStreamId encapsulates the retrieval of a new stream id and the addition to metrics as a reusable function
+// for the different mapper implementation
+func (sip *streamIdProcessor) getNewStreamId(streamId int16) (int16, error) {
+	var newId, err = sip.mapper.GetNewIdFor(streamId)
+	if err != nil {
+		return -1, err
+	}
+	if sip.metrics != nil {
+		sip.metrics.Add(1)
+	}
+	return newId, err
+}
+
+// releaseStreamId encapsulates the release of the synthetic stream id and the subtraction of the metrics
+// as a reusable function for the different mapper implementation
+func (sip *streamIdProcessor) releaseStreamId(syntheticId int16) (int16, error) {
+	var originalId, err = sip.mapper.ReleaseId(syntheticId)
+	if sip.metrics != nil && err != nil {
+		sip.metrics.Subtract(1)
+	}
+	return originalId, err
+}
```

Comment:
Move these 2 methods closer to the other processor methods. Currently there are 2 global functions in the middle of them.

Reply:
Done